### ISSUE TITLE
"Expected X, but got Y" exception when reading results

### ISSUE THREAD

**Issue by shift-evgeny** [neo4j/neo4j-dotnet-driver#83]
Using the latest code on the 1.0 branch I tried returning all nodes, like this:

```
session.Run("MATCH (n) RETURN n")
```

Every time I get an exception like this:

```
Expect 99, but got 44
   at Neo4j.Driver.Internal.Connector.ChunkedInputStream.ReadSpecifiedSize(Byte[] buffer) in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Connector\ChunkedInputStream.cs:line 131
   at Neo4j.Driver.Internal.Connector.ChunkedInputStream.Ensure(Int32 size) in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Connector\ChunkedInputStream.cs:line 113
   at Neo4j.Driver.Internal.Connector.ChunkedInputStream.ReadByte() in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Connector\ChunkedInputStream.cs:line 51
   at Neo4j.Driver.Internal.Packstream.PackStream.Unpacker.UnpackStructHeader() in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Packstream\PackStream.cs:line 556
   at Neo4j.Driver.Internal.Packstream.PackStreamMessageFormatV1.ReaderV1.Read(IMessageResponseHandler responseHandler) in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Packstream\PackStreamMessageFormatV1.cs:line 53
   at Neo4j.Driver.Internal.Connector.SocketClient.Receive(IMessageResponseHandler responseHandler) in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Connector\SocketClient.cs:line 107
```

(The numbers differ from run to run.)

**Comment by zhenlineo** [comment]
While I am finding a big database to test, could you help me to verify these questions:

Do you have the problem without this change https://github.com/neo4j/neo4j-dotnet-driver/pull/74, aka, with this line 
`var chunkSize = _bitConverter.ToUInt16(_headTailBuffer);`
 revert back to `var chunkSize = _bitConverter.ToInt16(_headTailBuffer);`
Do you have the problem with 1.0.0?

Thanks in advance.

**Comment by shift-evgeny** [comment]
Yes, the problem happens with 1.0.0. I get the same error if I change that line to `var chunkSize = _bitConverter.ToInt16(_headTailBuffer);`

**Comment by zhenlineo** [comment]
This might be a packstream bug. We have not seen it before.

May I ask if you data is not so sensitive that you could turn on logging and give us the debug log near the error?

So the debug trace would be something like:

```
[Trace] => C: 00 46 B1 01 D0 12 6E 65 6F 34 6A 2D 64 6F 74 6E 65 74 2F 31 2E 30 2E 30 A3 86 73 63 68 65 6D 65 85 62 61 73 69 63 89 70 72 69 6E 63 69 70 61 6C 85 6E 65 6F 34 6A 8B 63 72 65 64 65 6E 74 69 61 6C 73 85 6E 65 6F 34 6A 00 00 00 17 B2 10 D0 12 52 45 54 55 52 4E 20 32 20 61 73 20 4E 75 6D 62 65 72 A0 00 00 00 02 B0 3F 00 00
[Trace] => S: 00 03
[Trace] => S: B1 70 A0
[Debug] => S: [SUCCESS []]
[Trace] => S: 00 00
[Trace] => S: 00 12
[Trace] => S: B1 70 A1 86 66 69 65 6C 64 73 91 86 4E 75 6D 62 65 72
The thread 0x257c has exited with code 0 (0x0).
The thread 0x2578 has exited with code 0 (0x0).
The thread 0x2570 has exited with code 0 (0x0).
[Debug] => S: [SUCCESS [{fields : [Number]}]]
[Trace] => S: 00 00
[Trace] => S: 00 04
[Trace] => S: B1 71 91 02
[Debug] => S: [RECORD [2]]
[Trace] => S: 00 00
[Trace] => S: 00 0A
[Trace] => S: B1 70 A1 84 74 79 70 65 81 72
[Debug] => S: [SUCCESS [{type : r}]]
[Trace] => S: 00 00
```

The meaning of the few lines are explained after ->

```
[Trace] => S: 00 00 -> end of previous message
[Trace] => S: 00 0A -> size of the new message
[Trace] => S: B1 70 A1 84 74 79 70 65 81 72 -> the message content
[Debug] => S: [SUCCESS [{type : r}]] -> the plain text of the message
[Trace] => S: 00 00 -> message end
```

So if this leak of information is okay for you, would you please turn on logging like this example code:
https://github.com/neo4j/neo4j-dotnet-driver/blob/1.1/Neo4j.Driver/Neo4j.Driver.IntegrationTests/ConnectionIT.cs#L48

and give us the trace a few lines before the error message.

If this info is too sensitive to you, then just play with it. Maybe you might find the source of the error.

The logging is only printed into output window when I run code with `Debug` in VS.

**Comment by shift-evgeny** [comment]
OK, I've enabled logging. To do this I had to hack BaseOutLogger to make it public and then inherit from it myself to write the output via log4net. It generated a 422 MB log file, which would never have worked via Debug.WriteLine(). So - as a separate issue - this driver really needs better built-in logging support, i.e. something that can write to a file without such hacks.

Anyway, here's the end of the log:

> 2016-06-09 15:24:34 [DEBUG] TryNeo4j - [Debug] => S: [RECORD [Neo4j.Driver.Internal.Node]]
> 2016-06-09 15:24:34 [DEBUG] TryNeo4j - [Trace] => S: 00 00
> 2016-06-09 15:24:34 [DEBUG] TryNeo4j - [Trace] => S: 00 69
> 2016-06-09 15:24:34 [DEBUG] TryNeo4j - [Trace] => S: B1 71 91 B3 4E CA 00 0B A3 9E 91 86 45 6E 74 69 74 79 A5 8A 45 6E 74 69 74 79 54 79 70 65 86 50 65 72 73 6F 6E 88 45 6E 74 69 74 79 49 64 CB 00 00 00 02 54 1D 8F 3C 89 46 69 72 73 74 4E 61 6D 65 8A 69 61 6F 65 66 6C 72 74 79 6F 87 4F 72 67 4E 61 6D 65 80 88 4C 61 73 74 4E 61 6D 65 8A 65 79 74 77 6D 68 6F 6F 73 6F
> 2016-06-09 15:24:34 [DEBUG] TryNeo4j - [Debug] => S: [RECORD [Neo4j.Driver.Internal.Node]]
> 2016-06-09 15:24:34 [DEBUG] TryNeo4j - [Trace] => S: 00 00
> 2016-06-09 15:24:34 [DEBUG] TryNeo4j - [Trace] => S: 00 6A
> 2016-06-09 15:25:41 [DEBUG] TryNeo4j - [Error] => Unable to unpack message from server, connection has been terminated.[Neo4j.Driver.V1.Neo4jException: Expect 106, but got 83
>    at Neo4j.Driver.Internal.Connector.ChunkedInputStream.ReadSpecifiedSize(Byte[] buffer) in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Connector\ChunkedInputStream.cs:line 131
>    at Neo4j.Driver.Internal.Connector.ChunkedInputStream.Ensure(Int32 size) in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Connector\ChunkedInputStream.cs:line 113
>    at Neo4j.Driver.Internal.Connector.ChunkedInputStream.ReadByte() in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Connector\ChunkedInputStream.cs:line 51
>    at Neo4j.Driver.Internal.Packstream.PackStream.Unpacker.UnpackStructHeader() in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Packstream\PackStream.cs:line 556
>    at Neo4j.Driver.Internal.Packstream.PackStreamMessageFormatV1.ReaderV1.Read(IMessageResponseHandler responseHandler) in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Packstream\PackStreamMessageFormatV1.cs:line 53
>    at Neo4j.Driver.Internal.Connector.SocketClient.Receive(IMessageResponseHandler responseHandler) in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Connector\SocketClient.cs:line 101, []]
> 2016-06-09 15:25:42 [DEBUG] TryNeo4j - [Error] => Expect 106, but got 83[Neo4j.Driver.V1.Neo4jException: Expect 106, but got 83
>    at Neo4j.Driver.Internal.Connector.ChunkedInputStream.ReadSpecifiedSize(Byte[] buffer) in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Connector\ChunkedInputStream.cs:line 131
>    at Neo4j.Driver.Internal.Connector.ChunkedInputStream.Ensure(Int32 size) in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Connector\ChunkedInputStream.cs:line 113
>    at Neo4j.Driver.Internal.Connector.ChunkedInputStream.ReadByte() in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Connector\ChunkedInputStream.cs:line 51
>    at Neo4j.Driver.Internal.Packstream.PackStream.Unpacker.UnpackStructHeader() in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Packstream\PackStream.cs:line 556
>    at Neo4j.Driver.Internal.Packstream.PackStreamMessageFormatV1.ReaderV1.Read(IMessageResponseHandler responseHandler) in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Packstream\PackStreamMessageFormatV1.cs:line 53
>    at Neo4j.Driver.Internal.Connector.SocketClient.Receive(IMessageResponseHandler responseHandler) in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Connector\SocketClient.cs:line 107
>    at Neo4j.Driver.Internal.Connector.SocketClient.Send(IEnumerable`1 messages, IMessageResponseHandler responseHandler) in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Connector\SocketClient.cs:line 90
>    at Neo4j.Driver.Internal.Connector.SocketConnection.Sync() in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Connector\SocketConnection.cs:line 67
>    at Neo4j.Driver.Internal.Session.<>c__DisplayClass8_0.<Run>b__0() in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\Session.cs:line 83
>    at Neo4j.Driver.Internal.LoggerBase.TryExecute[T](Func`1 func) in C:\G it\neo4j-dotnet-driver\Neo4j.Driver\Neo4j.Driver\Internal\LoggerBase.cs:line 49, []]

**Comment by zhenlineo** [comment]
Thanks a lot!

Might be missing a while loop here:
https://github.com/neo4j/neo4j-dotnet-driver/blob/1.1/Neo4j.Driver/Neo4j.Driver/Internal/Connector/SocketExtensions.cs#L30

Working on tests to verify now.

P.S. The debug logging is not really supported officially. I am super happy you figured out how to inject the logging into your logging system! That's the design that we leave the logging hook and anyone who wants to inject the driver logging into their logging system would impl themselves. Though the BaseLogger is very handy to have publicly.

**Comment by zhenlineo** [comment]
I merged a fix, let me know if your problem get solved with the new change to 1.0

**Comment by shift-evgeny** [comment]
Thank you, I'm not longer getting the error with this fix.
